api/base/exceptions.py

```diff
@@ -26,28 +52,33 @@ def json_api_exception_handler(exc, context):
                 }
             ])
         elif isinstance(message, dict):
-            for error_key, error_description in message.iteritems():
-                if error_key in top_level_error_keys:
-                    errors.append({error_key: error_description})
-                elif error_key in resource_object_identifiers:
-                    if isinstance(error_description, basestring):
-                        error_description = [error_description]
-                    errors.extend([{'source': {'pointer': '/data/' + error_key}, 'detail': reason} for reason in error_description])
-                elif error_key == 'attributes':
-                    if isinstance(error_description, list):
-                        errors.extend([{'source': {'pointer': '/data/' + error_key}, 'detail': reason} for reason in error_description])
-                else:
-                    if isinstance(error_description, basestring):
-                        error_description = [error_description]
-                    errors.extend([{'source': {'pointer': '/data/attributes/' + error_key}, 'detail': reason} for reason in error_description])
-
+            errors.extend(dict_error_formatting(message))
         else:
             if isinstance(message, basestring):
                 message = [message]
-            errors.extend([{'detail': error} for error in message])
+            for error in message:
+                if isinstance(error, dict):
+                    errors.extend(dict_error_formatting(error))
+                else:
+                    errors.append({'detail': error})
 
         response.data = {'errors': errors}
 
+        # For bulk operations: If 400 error, return request data with response.
+        if response.status_code == 400 and "non_field_errors" not in message:
+            request_data = context['request'].data
+            if isinstance(request_data, list):
+                formatted_request_data = []
+                for data in request_data:
+                    formatted = {'type': data.pop('type')}
+                    id = data.pop('id')
+                    if id is not None:
+                        formatted['id'] = id
+                    formatted['attributes'] = data
+                    formatted_request_data.append(formatted)
+
+                response.data['meta'] = {'request_data': formatted_request_data}
```

Comment:
I am suggesting that the index of the invalid data be included in the pointer. In your example above, the pointer for the first error would be `/data/0/attributes/title` and the second pointer would be `/data/1/attributes/category`.

Not that we need to optimize our response payload sizes at this time, but including the request data seems extraneous; it doesn't give any information the client doesn't already have.